src/driver/xdma_sgdma.c

```diff
@@ -261,22 +149,59 @@ ssize_t xdma_channel_read_write(struct xdma_channel *chnl,
 		return -EINVAL;
 	}
 
-	rv = check_transfer_align(engine, buf, count, *pos, 1);
-	if (rv) {
-		pr_info("Invalid transfer alignment detected\n");
-		return rv;
-	}
+	if (sg_alloc_table(&sgt, SG_MAX_SINGLE_ALLOC, GFP_KERNEL))
+		return -ENOMEM;
 
-	memset(&cb, 0, sizeof(struct xdma_io_cb));
-	cb.buf = (char __user *)buf;
-	cb.len = count;
-	rv = xdma_channel_map_user_buf_to_sgl(&cb, write);
-	if (rv < 0)
-		return rv;
+	while (iov_iter_count(iter)) {
+		pages = (struct page **) sgt.sgl;
+		pages += SG_MAX_SINGLE_ALLOC * (PAGE_PTRS_PER_SGL - 1);
+
+		size = iov_iter_get_pages(iter, pages, LONG_MAX,
+				SG_MAX_SINGLE_ALLOC, &offset);
+		if (size < 0) {
+			res = size;
+			goto out;
+		}
+
+		rv = check_transfer_align(engine, (void *) offset, size, pos, 1);
+
+		if (rv) {
+			pr_info("Invalid transfer alignment detected\n");
+			res = rv;
+			goto out;
+		}
+
+		for (left = size, i = 0; left > 0; left -= len, i++) {
+			len = min_t(size_t, PAGE_SIZE - offset, left);
+			sg_set_page(&sgt.sgl[i], pages[i], len, offset);
+			offset = 0;
+		}
+		sgt.nents = pages_nr = i;
+		if (i < SG_MAX_SINGLE_ALLOC)
```

Comment:
Why not just always mark the end? Technically it's already done if  i==SG_MAX_SINGLE_ALLOC, but it doesn't hurt to do it again.

Reply:
`sgt.sgl[SG_MAX_SINGLE_ALLOC]` is a buffer overflow. But this wrong nonetheless, I should be setting `i - 1` as the end. And then I added a sg_unmark_end at the end of the while() so we can reuse the sgl.